src/pybullet_helpers/robots/single_arm.py

```diff
@@ -1,107 +1,179 @@
+        self.robot_id = p.loadURDF(
+            self.urdf_path(),
+            basePosition=self._base_pose.position,
+            baseOrientation=self._base_pose.quat_xyzw,
+            useFixedBase=True,
+            physicsClientId=self.physics_client_id,
+        )
 
-        Represents position control of the arm and finger joints.
-        """
-        return Box(np.array(self.joint_lower_limits, dtype=np.float32),
-                   np.array(self.joint_upper_limits, dtype=np.float32),
-                   dtype=np.float32)
+        # Robot initially at home pose
+        self.go_home()
 
     @classmethod
     @abc.abstractmethod
     def get_name(cls) -> str:
         """Get the name of the robot."""
         raise NotImplementedError("Override me!")
 
+    @classmethod
     @abc.abstractmethod
-    def _initialize(self) -> None:
+    def urdf_path(cls) -> str:
+        """Get the path to the URDF file for the robot."""
         raise NotImplementedError("Override me!")
 
+    @property
+    def action_space(self) -> Box:
+        """The action space for the robot.
+
+        Represents position control of the arm and finger joints.
+        """
+        return Box(
+            np.array(self.joint_lower_limits, dtype=np.float32),
+            np.array(self.joint_upper_limits, dtype=np.float32),
+            dtype=np.float32,
+        )
+
     @property
     @abc.abstractmethod
-    def robot_id(self) -> int:
-        """The PyBullet ID for the robot."""
+    def end_effector_name(self) -> str:
+        """The name of the end effector."""
         raise NotImplementedError("Override me!")
 
     @property
-    @abc.abstractmethod
     def end_effector_id(self) -> int:
-        """The PyBullet ID for the end effector."""
-        raise NotImplementedError("Override me!")
+        """The PyBullet joint ID for the end effector."""
+        return self.joint_names.index(self.end_effector_name)
 
     @property
     @abc.abstractmethod
-    def left_finger_id(self) -> int:
-        """The PyBullet ID for the left finger."""
+    def tool_link_name(self) -> str:
+        """The name of the end effector link (i.e., the tool link)."""
         raise NotImplementedError("Override me!")
 
+    @cached_property
+    def arm_joints(self) -> List[int]:
+        """The Pybullet joint IDs of the joints of the robot arm.
+
+        Note these are joint indices not body IDs.
+        """
+        joint_ids = get_kinematic_chain(
+            self.robot_id,
+            self.end_effector_id,
+            physics_client_id=self.physics_client_id,
+        )
+        # NOTE: pybullet tools assumes sorted arm joints.
+        joint_ids = sorted(joint_ids)
+        joint_ids.extend([self.left_finger_id, self.right_finger_id])
+        return joint_ids
+
+    @cached_property
+    def joint_infos(self) -> List[JointInfo]:
+        """Get the joint info for each joint of the robot."""
+        joint_ids = list(
```

Comment:
use `get_joints` here?

Reply:
There was some weird infinite recursion when I was trying to improve this late at night. Will figure out a better way